sw/device/lib/testing/test_framework/coverage_llvm.c

```diff
@@ -19,33 +19,49 @@
 int __llvm_profile_runtime;
 
 /**
- * Sends the given buffer as a hex string over UART.
+ * Coverage buffer in `.bss` to avoid stack issues.
  */
-static void send_buffer(uint8_t *buf, size_t len) {
-  for (uint32_t i = 0; i < len; ++i) {
-    base_printf("%02X", buf[i]);
-  }
-}
+static char buf[0x8000] = {0};
 
-void test_coverage_send_buffer(void) {
+/**
+ * Sends the given buffer as a hex string over UART.
+ *
+ * Lines are limited to 80 characters. Longer buffers will be broken to multiple
+ * lines.
+ */
+void coverage_send_buffer(void) {
   // It looks like we don't have a way to read the profile buffer incrementally.
   // Thus, we define the following buffer.
-  uint8_t buf[8192] = {0};
   // Write the profile buffer to the buffer defined above.
-  uint32_t buf_size = (uint32_t)__llvm_profile_get_size_for_buffer();
+  size_t buf_size = (size_t)__llvm_profile_get_size_for_buffer();
   if (buf_size > sizeof(buf)) {
     LOG_ERROR("ERROR: LLVM profile buffer is too large: %u bytes.", buf_size);
   } else {
-    __llvm_profile_write_buffer((char *)buf);
+    __llvm_profile_write_buffer(buf);
     // Send the buffer along with its length and CRC32.
-    base_printf("\r\nLLVM profile data (length: %u, CRC32: ", buf_size);
     uint32_t checksum = crc32(buf, buf_size);
-    send_buffer((uint8_t *)&checksum, sizeof(checksum));
-    base_printf("):\r\n");
-    send_buffer(buf, buf_size);
-    base_printf("\r\n");
+    LOG_INFO("LLVM profile data (length: %u bytes, CRC32: 0x%08x):", buf_size,
+             checksum);
+    enum {
+      kBytesPerLine = 40,
+      // Print one less byte in the first line to account for "0x".
+      kBytesFirstLine = kBytesPerLine - 1,
+    };
+    size_t bytes_this_line =
+        buf_size < kBytesFirstLine ? buf_size : kBytesFirstLine;
+    char *read_ptr = (char *)buf + buf_size - bytes_this_line;
+    LOG_INFO("0x%!x", kBytesFirstLine, read_ptr);
+    buf_size -= kBytesFirstLine;
```

Comment:
Oh closer inspection, I think lines 53 and 54 should use `bytes_per_line` instead of `kBytesFirstLine`.

Reply:
Good catch! Fixed. LLVM profile buffers always end up larger than ~100 bytes even for really small test programs but regardless we must handle this case correctly.